### PR TITLE
YETI-3635: Add support for single-disk encryption on GCP

### PR DESCRIPTION
**Pull request by xcllnt**
Single-disk encryption is a way of encrypting whereby the encrypted image has the same number of disks as the unencrypted/stock image. While currently we assume there's only a root disk, the mechanism extends to images with multiple disks.

The immediate upshot is that bracketized images can be launched from the GCP or Azure console and that we therefore don't need to use the launch command in the CLI. There are less immediate advantages as well on AWS and ESX, which is why single-disk encryption is not specific to any CSP, but rather the new way to encrypt across all platforms.

Globally:
 o  Introduce METAVISOR_DISK_SIZE. It holds the size of the
    metavisor disk.
 o  Add options --single-disk and --no-single-disk
    Provide a reasonable default when neither is specified.
    Add 'single-disk' to the bracket configuration for GCP.
 o  Add encryptor_did_single_disk()

For GCP:
 o  Support encrypting legacy and single-disk on GCP.
 o  Support updating legacy and single-disk on GCP

For AWS and ESX:
 o  Remove single_disk from the bracket configuration. They'll
    support single-disk encryption in a follow-up commit to
    minimize risk.

While here:
 o  Remove redundant handling of crypto_policy_type

Jira: YETI-3635

### CODE REVIEW

File: brkt_cli/aws/encrypt_ami.py [hunk 1]
```diff
@@ -118,7 +118,6 @@ def _run_encryptor_instance(
 
     if instance_config is None:
         instance_config = InstanceConfig()
-    instance_config.brkt_config['crypto_policy_type'] = crypto_policy
```
Comment (krishnanrs) [inline brkt_cli/aws/encrypt_ami.py]: I believe we still need to pass this value in the instance_config to the encryptor. I am not sure if the metavisor defaults to XTS if this is not present, however in case the user specifies a different crypto policy using the command line args, this field would be required.

File: brkt_cli/gcp/encrypt_gcp_image.py [hunk 2]
```diff
@@ -4,178 +4,152 @@
...
+        gcp_svc.wait_for_detach(values.zone, instance_name)
+        guest_disk_size = gcp_svc.get_disk_size(values.zone, instance_name)
+
+        # Create a blank disk. The size of which depends on the crypto
+        # policy and whether we do single-disk encryption or not.
+        encrypted_disk_size = guest_disk_size
+        if values.crypto == CRYPTO_GCM:
+            encrypted_disk_size *= 2
+        if values.single_disk:
+            encrypted_disk_size += METAVISOR_DISK_SIZE
```
Comment (krishnanrs) [inline brkt_cli/gcp/encrypt_gcp_image.py]: Shouldn't this be METAVISOR_DISK_SIZE + 1 to accommodate for the encryption headers?

**Comment by bburtin** [comment]
Can you please update the commit comment with some information about what single disk encryption is, and why we're making this change?  It would really help reviewers understand the motivation of the work, especially if they've been (ahem) out of the loop for a while.

File: brkt_cli/gcp/encrypt_gcp_image.py [hunk 3]
```diff
@@ -4,178 +4,152 @@
...
-        log.info("Creating metavisor image")
-        gcp_svc.create_gcp_image_from_disk(zone, encrypted_image_name, encryptor)
+        if not values.single_disk:
+            # Snapshot the encrypted guest disk
+            log.info("Creating snapshot of encrypted image disk")
+            gcp_svc.create_snapshot(values.zone, encrypted_image_disk,
+                                    encrypted_image_name)
+            root_disk = encryptor
+        else:
+            root_disk = encrypted_image_disk
```
Comment (krishnanrs) [inline brkt_cli/gcp/encrypt_gcp_image.py]: So is the actual magic of creating a MV and guest partition in the single disk being handled by the MV/encryptor code?

File: brkt_cli/encryptor_service.py [hunk 4]
```diff
@@ -260,7 +250,14 @@ def wait_for_encryption(enc_svc,
     raise EncryptionError('Encryption service unavailable')
 
 
+def encryptor_did_single_disk(enc_svc):
+    sd = enc_svc.fetch('/single_disk')
```
Comment (bburtin) [inline brkt_cli/encryptor_service.py]: Do we really want an API endpoint that returns `text/plain`?  Would it make sense to add some sort of Metavisor properties and point and have single-disk encryption be a field in the response?  That way we don't need to add a new endpoint for every property we add in the future.

File: brkt_cli/gcp/__init__.py [hunk 5]
```diff
@@ -50,6 +53,19 @@ def run_encrypt(values, config):
     if not values.verbose:
         logging.getLogger('googleapiclient').setLevel(logging.ERROR)
 
+    if not values.encryptor_image:
```
Comment (bburtin) [inline brkt_cli/gcp/__init__.py]: Should this work be in a separate commit?  It's not immediately obvious how it relates to single-disk encryption.

File: brkt_cli/gcp/__init__.py [hunk 6]
```diff
@@ -60,25 +76,15 @@ def run_encrypt(values, config):
         brkt_env=brkt_env,
         launch_token=lt
     )
+
     encrypted_image_id = encrypt_gcp_image.encrypt(
         gcp_svc=gcp_svc,
         enc_svc_cls=encryptor_service.EncryptorService,
-        image_id=values.image,
-        encryptor_image=values.encryptor_image,
+        values=values,
```
Comment (bburtin) [inline brkt_cli/gcp/__init__.py]: I'm not sure we want to go in this direction.  I understand that it's painful to have 20 arguments to a function, but passing in an unstructured dictionary is also not great.  Someone reading the `encrypt()` source won't know which values are being passed in.  If you want to encapsulate the arguments, you could add an `EncryptValues` class or something similar.

File: brkt_cli/aws/encrypt_ami.py (hunk 1 above)
Comment (xcllnt) [reply brkt_cli/aws/encrypt_ami.py]: Common code now sets crypto_policy_type. See instance_config_args.py.

File: brkt_cli/gcp/encrypt_gcp_image.py (hunk 2 above)
Comment (xcllnt) [reply brkt_cli/gcp/encrypt_gcp_image.py]: No need. There's enough unallocated space on the MV disk to cover the overhead.

File: brkt_cli/gcp/encrypt_gcp_image.py (hunk 3 above)
Comment (xcllnt) [reply brkt_cli/gcp/encrypt_gcp_image.py]: Yes. The prime responsibility of the CLI is to size the empty disk correctly and to construct the final image with the right disks. The metavisor will do the rest.

File: brkt_cli/encryptor_service.py (hunk 4 above)
Comment (xcllnt) [reply brkt_cli/encryptor_service.py]: Why not? There's nothing wrong with another endpoint. And since this is temporary, it's actually better not to add it to the status. This is especially true given how the metavisor is implemented.

File: brkt_cli/gcp/__init__.py (hunk 5 above)
Comment (xcllnt) [reply brkt_cli/gcp/__init__.py]: It's part of the same same restructuring of eliminating a zillion arguments and only passing the values. By moving this up one level, we have a clear separation between filling in the blanks at the top-most CSP-specific action and read-only use further down the stack.

File: brkt_cli/gcp/__init__.py (hunk 6 above)
Comment (xcllnt) [reply brkt_cli/gcp/__init__.py]: It's not merely painful, it's actually worse. Adding an extra argument was beyond my capabilities.

Note that arguments are actually a lot more unstructured than a name space, especially if you start providing default arguments. Looking at a function you simply won't know where values are coming from, which means that understanding the code is a lot harder than it has to be.

Passing values as-is, makes it a lot clearer that they're effectively coming from the command-line with some post-processing. Moving that post-processing, as high up the stack as possible and treating values as read-only as much as possible, makes the logic a lot more understandable.

Not to mention that a function call with 20 arguments is not readable at all.

I think it's a huge improvement.

Comment (bburtin) [reply brkt_cli/gcp/__init__.py]: What about the idea of adding a new class that encapsulates the fields that are passed to the `encrypt()` function?  It's brittle when application code depends on the structure of the command line spec.  Having a separate class is a little more tedious, but I think it's less painful than tightly coupling the application code to the format of the command line arguments.

Comment (xcllnt) [reply brkt_cli/gcp/__init__.py]: It's already a class. values is a proper object. There's no dependency on command-line order or its structure. Hence it's no more brittle than creating an encapsulating class. What we have is less obfuscating, which in my eyes means it's preferred. So, once we have a good enough reason to abstract (such as when we want to move more CSP-specific code to common code to reduce the duplication and find that the ad-hoc and different naming of fields in values makes that painful), then I'd say "yeah, maybe" (I would first simply rename the fields to remove the gratuitous differences :-)
